ruma-identifiers/src/mxc_uri.rs

```diff
@@ -1,75 +1,77 @@
-//! Matrix-spec compliant mxc:// urls.
+//! A URI that should be a Matrix-spec compliant MXC URI.
 
-use std::{convert::TryFrom, fmt, str::FromStr};
+use std::{convert::TryFrom, fmt, num::NonZeroU8};
 
 use ruma_identifiers_validation::mxc_uri::validate;
 
-use crate::{Error, ServerName, ServerNameBox};
+use crate::ServerName;
 
-/// Matrix-spec compliant mxc:// urls.
+/// A URI that should be a Matrix-spec compliant MXC URI.
 #[derive(Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
 pub struct MxcUri {
-    server_name: ServerNameBox,
-    media_id: Box<str>,
+    full_uri: Box<str>,
+    slash_idx: Option<NonZeroU8>,
 }
 
 impl MxcUri {
-    /// Returns the media ID of this mxc://.
-    pub fn media_id(&self) -> &str {
-        &self.media_id
+    /// If this is a valid MXC URI, returns the media ID.
+    pub fn media_id(&self) -> Option<&str> {
+        self.slash_idx.map(|idx| &self.full_uri[idx.get() as usize + 1..])
     }
 
-    /// Returns the server name of this mxc://.
-    pub fn server_name(&self) -> &ServerName {
-        &self.server_name
+    /// If this is a valid MXC URI, returns the server name.
+    pub fn server_name(&self) -> Option<&ServerName> {
+        self.slash_idx
+            .map(|idx| <&ServerName>::try_from(&self.full_uri[6..idx.get() as usize]).unwrap())
     }
 
-    fn mxc_uri_fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
-        f.write_fmt(format_args!("mxc://{}/{}", self.server_name, self.media_id))
+    /// If this is a valid MXC URI, returns a `(server_name, media_id)` tuple.
+    pub fn parts(&self) -> Option<(&ServerName, &str)> {
+        self.slash_idx.map(|_| (self.server_name().unwrap(), self.media_id().unwrap()))
```

Comment:
Can you implement server_name and media_id in terms of parts rather than the lithe way around? Then we don't need so many unwrap's.